File: CS/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
#region #usings
using DevExpress.XtraRichEdit.API.Native;
using DevExpress.XtraRichEdit.Layout.Export;
using System.Drawing.Drawing2D;
#endregion #usings

namespace CustomMarks
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();

            richEditControl1.LoadDocument("Search.rtf");

        }

        #region #custommark
        private void btn_Mark_Click(object sender, EventArgs e)
        {
            Document doc = richEditControl1.Document;
            CustomMark m = doc.CreateCustomMark(doc.Selection.Start, Color.DarkOrange);   
        }

        private void richEditControl1_CustomMarkDraw(object sender, DevExpress.XtraRichEdit.RichEditCustomMarkDrawEventArgs e)
        {        
            foreach (CustomMarkVisualInfo info in e.VisualInfoCollection)
            {
                Document doc = richEditControl1.Document;
                CustomMark mark = doc.GetCustomMarkByVisualInfo(info);
                e.Graphics.SmoothingMode = SmoothingMode.AntiAlias;

                Color curColor = (Color)info.UserData;
                if (mark.Position < doc.Selection.Start) curColor = Color.Green;
                    Pen p = new Pen(curColor, 3);
                    p.StartCap = LineCap.Flat;
                    p.EndCap = LineCap.ArrowAnchor;
                    e.Graphics.DrawLine(p, new Point(0 , info.Bounds.Y), info.Bounds.Location);
            }
        }
        #endregion #custommark

        private void btn_DeleteMarks_Click(object sender, EventArgs e)
        {
            Document doc = richEditControl1.Document;
            for (int i = 0; i < doc.CustomMarks.Count; doc.DeleteCustomMark(doc.CustomMarks[i])) ;
        }
    }
}

File: VB/Form1.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
#Region "#usings"
Imports DevExpress.XtraRichEdit.API.Native
Imports DevExpress.XtraRichEdit.Layout.Export
Imports System.Drawing.Drawing2D
#End Region ' #usings

Namespace CustomMarks
	Partial Public Class Form1
		Inherits Form

		Public Sub New()
			InitializeComponent()

			richEditControl1.LoadDocument("Search.rtf")

		End Sub

		#Region "#custommark"
		Private Sub btn_Mark_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btn_Mark.Click
			Dim doc As Document = richEditControl1.Document
			Dim m As CustomMark = doc.CustomMarks.Create(doc.Selection.Start, Color.DarkOrange)
		End Sub

		Private Sub richEditControl1_CustomMarkDraw(ByVal sender As Object, ByVal e As DevExpress.XtraRichEdit.RichEditCustomMarkDrawEventArgs) Handles richEditControl1.CustomMarkDraw
			For Each info As CustomMarkVisualInfo In e.VisualInfoCollection
				Dim doc As Document = richEditControl1.Document
				Dim mark As CustomMark = doc.CustomMarks.GetByVisualInfo(info)
				e.Graphics.SmoothingMode = SmoothingMode.AntiAlias

				Dim curColor As Color = CType(info.UserData, Color)
				If mark.Position < doc.Selection.Start Then
					curColor = Color.Green
				End If
				Using p As New Pen(curColor, 3)
					p.StartCap = LineCap.Flat
					p.EndCap = LineCap.ArrowAnchor
					e.Graphics.DrawLine(p, New Point(0, info.Bounds.Y), info.Bounds.Location)
				End Using
			Next info
		End Sub
		#End Region ' #custommark

		Private Sub btn_DeleteMarks_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btn_DeleteMarks.Click
			Dim doc As Document = richEditControl1.Document
			Dim i As Integer = 0
			Do While i < doc.CustomMarks.Count

				doc.CustomMarks.Remove(doc.CustomMarks(i))
			Loop
		End Sub
	End Class
End Namespace